### DISCUSSION TITLE
Creating KTX2 in wasm

### DISCUSSION THREAD

**Discussion by jocabola** [KhronosGroup/KTX-Software#685] [Q&A] [answered]
Is it possible to create a KTX2 texture from a source JPG or PNG using the libktx wasm in web? Or do I need to wrap the `toktx` command in node or other server-side technology to generate them? It seems like client machines will be usually more powerful and would make sense to do so, but I am stuck using the wasm module right now....

Thanks!

**Comment by jocabola** [comment]
Hello again,

Sorry to insist but anybody knows the answer to this? Is the `toktx` functionality exposed in the wasm lib or is this only intended for transcoding currently?

If KTX generation is possible with the wasm lib is there any documentation out there? I couldn't see any `ktxTexture_Create` or similar exposed in the JS lib.

Thanks,
Edu

**Comment by MarkCallow** [comment]
Sorry for the delay answering.

`libktx.{js,wasm}` is based on `ktx_read`, i.e the read only library. That is because that is all the wrapper currently supports. It is possible to compile the full libktx including encoders (not 100% sure about the ASTC encoder, definitely with the Basisu encoders) with Emscripten. There are two ways to use it:

1. Extend `interface/js_binding/ktx_wrapper.cpp` to support write functions and modify the wasm build to build the full library.
2. Write an app in c++, compile it too with Emscripten, and link it with the full libktx.

Because `toktx` is a command line program with no graphical UI, and it reads and writes files, I haven't seen any point compiling it to run in a browser as I expect quite a bit of further work is necessary make it usable. `tests/loadtests/gl3loadtests` can be compiled with Emscripten (as `es3loadtests`) and includes a test that runs the BasisU encoder. It runs in a browser.

I'd happily accept PRs to extend `ktx_wrapper` or make `toktx` usable in a browser or both. If you are tempted to do the latter, you should hold off for a month. `toktx` will be replaced soon with a new tool.

**Reply by jocabola** [reply]
Thank you so much for the clear answer. We have a temporary workaround for now. Will keep an eye on the upcoming new tool! Thank you!

**Reply by donmccurdy** [reply]
Is this answer still accurate today, with the v4.3 alpha release?

**Reply by MarkCallow** [reply]
Yes it is still accurate.

Work has been done to add a small number of create-related functions to the wrapper as part of a soon to be released project. As the work only adds the minimal functions needed by that project, the updated wrapper will remain as part of that project.

As I said before, I'd happily accept a PR to extend the wrapper to the full functionality of libktx.

**Comment by MarkCallow** [comment, marked as answer]
As of v4.4.0 the JS wrapper provides the full functionality of libktx. Download https://github.com/KhronosGroup/KTX-Software/releases/download/v4.4.0/KTX-Software-4.4.0-Web-libktx.zip. Use https://github.com/KhronosGroup/KTX-Software/releases/download/v4.4.0/KTX-Software-4.4.0-Web-libktx_read.zip if you only need read functionality.
